edge-modules/api-proxy-module/src/monitors/certs_monitor.rs

```diff
@@ -0,0 +1,327 @@
+) -> Result<(JoinHandle<Result<()>>, ShutdownHandle), Error> {
+    const NGINX_CERTIFICATE_MONITOR_POLL_INTERVAL_SECS: tokio::time::Duration =
+        tokio::time::Duration::from_secs(1);
+
+    let shutdown_signal = Arc::new(tokio::sync::Notify::new());
+    let shutdown_handle = ShutdownHandle(shutdown_signal.clone());
+
+    let module_id = env::var("IOTEDGE_MODULEID")
+        .context(format!("Missing env var {:?}", "IOTEDGE_MODULEID"))?;
+    let generation_id = env::var("IOTEDGE_MODULEGENERATIONID").context(format!(
+        "Missing env var {:?}",
+        "IOTEDGE_MODULEGENERATIONID"
+    ))?;
+    let gateway_hostname = env::var("IOTEDGE_GATEWAYHOSTNAME")
+        .context(format!("Missing env var {:?}", "IOTEDGE_GATEWAYHOSTNAME"))?;
+    let workload_url = env::var("IOTEDGE_WORKLOADURI")
+        .context(format!("Missing env var {:?}", "IOTEDGE_WORKLOADURI"))?;
+    let mut cert_monitor = CertificateMonitor::new(
+        module_id,
+        generation_id,
+        gateway_hostname,
+        workload_url,
+        Duration::days(PROXY_SERVER_CERT_VALIDITY_DAYS),
+    )
+    .context("Could not create cert monitor client")?;
+
+    let monitor_loop: JoinHandle<Result<()>> = tokio::spawn(async move {
+        loop {
+            let wait_shutdown = shutdown_signal.notified();
+            futures::pin_mut!(wait_shutdown);
+
+            match futures::future::select(
+                time::delay_for(NGINX_CERTIFICATE_MONITOR_POLL_INTERVAL_SECS),
+                wait_shutdown,
+            )
+            .await
+            {
+                Either::Right(_) => {
+                    log::warn!("Shutting down certs monitor!");
+                    return Ok(());
+                }
+                Either::Left(_) => {}
+            };
+
+            //If root cert has rotated, we need to notify the watchdog to restart nginx.
+            let mut need_notify = false;
+            //Check for rotation. If rotated then we notify.
+            match cert_monitor.has_bundle_of_trust_rotated().await {
+                Ok((has_bundle_of_trust_rotated, trust_bundle)) => {
+                    if has_bundle_of_trust_rotated == true {
+                        //If we have a new cert, we need to write it in file system
+                        let result = utils::write_binary_to_file(
+                            trust_bundle.as_bytes(),
+                            PROXY_SERVER_TRUSTED_CA_PATH,
+                        );
+                        match result {
+                            Err(err) => panic!("{:?}", err),
+                            Ok(_) => (),
+                        }
+                        need_notify = true;
+                    }
+                }
+                Err(err) => log::error!("Error while trying to get trust bundle {:?}", err),
+            };
+
+            //Same thing as above but for private key and server cert
+            match cert_monitor.need_to_rotate_server_cert(Utc::now()).await {
```

Comment:
same comments as above

Reply:
Why on this one? I don't need to convert it?

Reply:
fixed